### ISSUE TITLE
Suggestion: don't issue error on property assignments to aliased functions

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#36330]
## Suggestion

```ts
type FooWithout = (n: number) => number;
type Foo = FooWithout & {
  bar: string;
};

() => {
  const foo: Foo = n => n; // ✅ error as expected because we forgot to assign property
  // foo.bar = '';
};

() => {
  const foo: Foo = n => n; // ✅ no error
  foo.bar = '';
};

() => {
  const fooWithout: FooWithout = n => n;
  const foo: Foo = fooWithout; // ❌ unexpected error, we didn't forget to assign property
  foo.bar = '';
};
```

## Use Cases

I am creating a function dynamically and I want to ensure it has a property assigned to it:

```ts
declare const createFooWithout: () => FooWithout;

() => {
  const foo: Foo = createFooWithout(); // ❌ unexpected error, we didn't forget to assign property
  foo.bar = '';
};
```

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by AnyhowStep** [comment]
This makes me uncomfortable and can lead to all kinds of unsound code.

```ts
type Foo2 = FooWithout & {
  bar: number;
};

  const fooWithout: FooWithout = n => n;
  const foo: Foo = fooWithout;
  foo.bar = '';

  const foo2 : Foo2 = fooWithout;
  foo2.bar = 32;

  console.log(foo.bar); //32
```

**Comment by OliverJAsh** [comment]
That's a good point…
